File: MyMcGill Mobile/src/ca/appvelopers/mcgillmobile/activity/ScheduleActivity.java
package ca.appvelopers.mcgillmobile.activity;

import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.joda.time.LocalTime;
import org.joda.time.Minutes;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import ca.appvelopers.mcgillmobile.App;
import ca.appvelopers.mcgillmobile.R;
import ca.appvelopers.mcgillmobile.activity.drawer.DrawerFragmentActivity;
import ca.appvelopers.mcgillmobile.activity.walkthrough.WalkthroughActivity;
import ca.appvelopers.mcgillmobile.fragment.DayFragment;
import ca.appvelopers.mcgillmobile.object.ClassItem;
import ca.appvelopers.mcgillmobile.object.Day;
import ca.appvelopers.mcgillmobile.object.Term;
import ca.appvelopers.mcgillmobile.util.Constants;
import ca.appvelopers.mcgillmobile.util.GoogleAnalytics;
import ca.appvelopers.mcgillmobile.util.Help;
import ca.appvelopers.mcgillmobile.util.Load;
import ca.appvelopers.mcgillmobile.util.Save;
import ca.appvelopers.mcgillmobile.util.downloader.ClassDownloader;

/**
 * @author Nhat-Quang Dao
 * Date: 22/01/14, 9:07 PM
 * 
 * This Activity loads the schedule from https://horizon.mcgill.ca/pban1/bwskfshd.P_CrseSchd
 */
public class ScheduleActivity extends DrawerFragmentActivity {
    private static final int CHANGE_SEMESTER_CODE = 100;

    private List<ClassItem> mClassList;
    private ViewPager mPager;
    private FragmentManager mSupportFragmentManager;
    private Term mTerm;

    @Override
	public void onCreate(Bundle savedInstanceState) {
        requestWindowFeature(Window.FEATURE_INDETERMINATE_PROGRESS);
        setContentView(R.layout.activity_schedule);
        super.onCreate(savedInstanceState);

        GoogleAnalytics.sendScreen(this, "Schedule");

        //Get the semester
        mTerm = App.getDefaultTerm();

        mClassList = new ArrayList<ClassItem>();

        //ViewPager stuff
        mSupportFragmentManager = getSupportFragmentManager();
        mPager = (ViewPager)findViewById(R.id.pager);

        //Load the stored info
        loadInfo();

        //Set up the class downloader
        executeClassDownloader();

        //Check if this is the first time the user is using the app
        if(Load.isFirstOpen(this)){
            //Show him the walkthrough if it is
            startActivity(new Intent(this, WalkthroughActivity.class));
            //Save the fact that the walkthrough has been seen at least once
            Save.saveFirstOpen(this);
        }
    }

    //Method that returns a list of courses for a given day
    public List<ClassItem> getClassesForDay(Day day){
        List<ClassItem> courses = new ArrayList<ClassItem>();

        //Go through the list of courses, find which ones have the same day
        for(ClassItem course : mClassList){
            if(course.getDays().contains(day)){
                courses.add(course);
            }
        }
        return courses;
    }


    @Override
    public void onConfigurationChanged(Configuration newConfig){
        super.onConfigurationChanged(newConfig);
        if(newConfig.orientation == Configuration.ORIENTATION_LANDSCAPE){
            setContentView(R.layout.activity_schedule_land);
            LayoutInflater inflater = (LayoutInflater)getSystemService(LAYOUT_INFLATER_SERVICE);

            LinearLayout timetableContainer = (LinearLayout) findViewById(R.id.timetable_container);
            fillTimetable(inflater, timetableContainer);

            LinearLayout scheduleContainer = (LinearLayout)findViewById(R.id.schedule_container);
            for(int i = 0; i < 7; i ++){
                LinearLayout coursesLayout = new LinearLayout(this);
                coursesLayout.setOrientation(LinearLayout.VERTICAL);
                coursesLayout.setLayoutParams(new LinearLayout.LayoutParams(
                        getResources().getDimensionPixelSize(R.dimen.cell_landscape_width),
                        ViewGroup.LayoutParams.WRAP_CONTENT));
                fillSchedule(inflater, coursesLayout, Day.getDay(i));
                scheduleContainer.addView(coursesLayout);

                //Line
                View line = new View(this);
                line.setBackgroundColor(getResources().getColor(R.color.black));
                line.setLayoutParams(new ViewGroup.LayoutParams(getResources().getDimensionPixelSize(R.dimen.line),
                        ViewGroup.LayoutParams.MATCH_PARENT));
                scheduleContainer.addView(line);
            }


        }
        else{
            setContentView(R.layout.activity_schedule);
            mPager = (ViewPager)findViewById(R.id.pager);
            loadInfo();
        }

        loadDrawer();
    }

    private void loadInfo(){
        //Title
        setTitle(mTerm.toString(this));

        //Clear the current course list
        mClassList.clear();
        for(ClassItem classItem : App.getClasses()){
            if(classItem.getTerm().equals(mTerm)){
                mClassList.add(classItem);
            }
        }

        SchedulePagerAdapter adapter = new SchedulePagerAdapter(mSupportFragmentManager);
        mPager.setAdapter(adapter);

        //Open it to the right day
        int day = Calendar.getInstance().get(Calendar.DAY_OF_WEEK);

        switch (day){
	        case Calendar.MONDAY:
	            mPager.setCurrentItem(500003);
	            break;
	        case Calendar.TUESDAY:
	            mPager.setCurrentItem(500004);
	            break;
	        case Calendar.WEDNESDAY:
	            mPager.setCurrentItem(500005);
	            break;
	        case Calendar.THURSDAY:
	            mPager.setCurrentItem(500006);
	            break;
	        case Calendar.FRIDAY:
	            mPager.setCurrentItem(500007);
	            break;
	        case Calendar.SATURDAY:
	            mPager.setCurrentItem(500008);
	            break;
	        case Calendar.SUNDAY:
	            mPager.setCurrentItem(500009);
	            break;
        }
    }

    private void fillTimetable(LayoutInflater inflater, LinearLayout timetableContainer){
        //Empty view for the days
        //Day name
        View dayView = inflater.inflate(R.layout.fragment_day_name, null);

        //Black line
        View dayViewLine = dayView.findViewById(R.id.day_line);
        dayViewLine.setVisibility(View.VISIBLE);

        timetableContainer.addView(dayView);

        //Cycle through the hours
        for(int hour = 8; hour < 22; hour++){
            //Start inflating a timetable cell
            View timetableCell = inflater.inflate(R.layout.item_day_timetable, null);

            //Quick check
            assert(timetableCell != null);

            //Put the correct time
            TextView time = (TextView)timetableCell.findViewById(R.id.cell_time);
            time.setText(Help.getShortTimeString(this, hour));

            //Add it to the right container
            timetableContainer.addView(timetableCell);
        }
    }

    //Method that fills the schedule based on given data
    private void fillSchedule(LayoutInflater inflater,
                              LinearLayout scheduleContainer, Day currentDay){
        //This will be used of an end time of a course when it is added to the schedule container
        LocalTime currentCourseEndTime = null;

        List<ClassItem> classItems = getClassesForDay(currentDay);

        //Day name
        View dayView = inflater.inflate(R.layout.fragment_day_name, null);
        TextView dayViewTitle = (TextView)dayView.findViewById(R.id.day_name);
        dayViewTitle.setText(currentDay.getDayString(this));

        scheduleContainer.addView(dayView);

        //Cycle through the hours
        for(int hour = 8; hour < 22; hour++){
            //Cycle through the half hours
            for(int min = 0; min < 31; min+= 30){
                //Initialize the current course to null
                ClassItem currentClass = null;

                //Get the current time
                LocalTime currentTime = new LocalTime(hour, min);

                //if currentCourseEndTime = null (no course is being added) or it is equal to
                //the current time in min (end of a course being added) we need to add a new view
                if(currentCourseEndTime == null || currentCourseEndTime.equals(currentTime)){
                    //Reset currentCourseEndTime
                    currentCourseEndTime = null;

                    //Check if there is a course at this time
                    for(ClassItem course : classItems){
                        //If there is, set the current course to that time, and calculate the
                        //ending time of this course
                        if(course.getStartTime().equals(currentTime)){
                            currentClass = course;
                            currentCourseEndTime = course.getEndTime();
                            break;
                        }
                    }

                    View scheduleCell;

                    //There is a course at this time
                    if(currentClass != null){
                        //Inflate the right view
                        scheduleCell = inflater.inflate(R.layout.item_day_class, null);

                        //Quick check
                        assert(scheduleCell != null);

                        //Set up all of the info
                        TextView courseName = (TextView)scheduleCell.findViewById(R.id.course_code);
                        courseName.setText(currentClass.getCourseCode());

                        TextView courseType = (TextView)scheduleCell.findViewById(R.id.course_type);
                        courseType.setText(currentClass.getSectionType());

                        TextView  courseTime = (TextView)scheduleCell.findViewById(R.id.course_time);
                        courseTime.setText(currentClass.getTimeString(this));

                        TextView courseLocation = (TextView)scheduleCell.findViewById(R.id.course_location);
                        courseLocation.setText(currentClass.getLocation());

                        //Find out how long this course is in terms of blocks of 30 min
                        int length = Minutes.minutesBetween(currentClass.getStartTime(), currentClass.getEndTime()).getMinutes() / 30;

                        //Set the height of the view depending on this height
                        LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT,
                                (int) this.getResources().getDimension(R.dimen.cell_30min_height) * length);
                        scheduleCell.setLayoutParams(lp);

                        //OnClick: CourseActivity (for                                                                                                                                                                                                  b                                                                                                                                               a detailed description of the course)
                        scheduleCell.setClickable(false);
                    }
                    else{
                        //Inflate the empty view
                        scheduleCell = inflater.inflate(R.layout.item_day_empty, null);

                        //Quick check
                        assert(scheduleCell != null);
                    }

                    //Add the given view to the schedule container
                    scheduleContainer.addView(scheduleCell);
                }
            }
        }
    }

    public void executeClassDownloader(){
        new ClassDownloader(this, mTerm){
            @Override
            protected void onPreExecute(){
                //Show the user we are refreshing
                setProgressBarIndeterminateVisibility(true);
            }

            // onPostExecute displays the results of the AsyncTask.
            @Override
            protected void onPostExecute(Boolean loadInfo) {
                if(loadInfo){
                    //Reload the adapter
                    loadInfo();
                }

                setProgressBarIndeterminateVisibility(false);
            }
        }.execute();
    }

    private class SchedulePagerAdapter extends FragmentStatePagerAdapter{
        public SchedulePagerAdapter(FragmentManager fm){
            super(fm);
        }

        @Override
        public Fragment getItem(int i) {
            Day currentDay = Day.getDay(i%7);
            return DayFragment.newInstance(currentDay);
        }

        @Override
        public int getCount() {
            return 1000000;
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu){
        getMenuInflater().inflate(R.menu.refresh, menu);

    	// change semester menu item
    	menu.add(Menu.NONE, Constants.MENU_ITEM_CHANGE_SEMESTER, Menu.NONE, R.string.schedule_change_semester);
    	return true;
    }
    
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            // Opens the context menu    
            case Constants.MENU_ITEM_CHANGE_SEMESTER:
            	Intent intent = new Intent(this, ChangeSemesterActivity.class);
                intent.putExtra(Constants.REGISTER_TERMS, false);
                intent.putExtra(Constants.TERM, mTerm);
                startActivityForResult(intent, CHANGE_SEMESTER_CODE);
            	return true;
            case R.id.action_refresh:
                //Start thread to retrieve schedule
                executeClassDownloader();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data){
        if(requestCode == CHANGE_SEMESTER_CODE){
            if(resultCode == RESULT_OK){
                //Get the chosen term
                mTerm = (Term)data.getSerializableExtra(Constants.TERM);
                executeClassDownloader();
            }
        }
        else{
            super.onActivityResult(requestCode, resultCode, data);
        }
    }
}

File: app/src/main/java/ui/ScheduleActivity.kt
/*
 * Copyright 2014-2022 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.guerinet.mymartlet.ui

import android.content.res.Configuration
import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import androidx.core.view.isVisible
import androidx.viewpager.widget.ViewPager
import com.guerinet.morf.Morf
import com.guerinet.morf.morf
import com.guerinet.mymartlet.R
import com.guerinet.mymartlet.model.Course
import com.guerinet.mymartlet.model.LocalTime
import com.guerinet.mymartlet.model.Term
import com.guerinet.mymartlet.model.atTime
import com.guerinet.mymartlet.model.place.Place
import com.guerinet.mymartlet.model.time
import com.guerinet.mymartlet.ui.dialog.list.TermDialogHelper
import com.guerinet.mymartlet.ui.walkthrough.WalkthroughActivity
import com.guerinet.mymartlet.util.Constants
import com.guerinet.mymartlet.util.DayUtils
import com.guerinet.mymartlet.util.Prefs
import com.guerinet.mymartlet.util.extensions.intentFor
import com.guerinet.mymartlet.util.extensions.start
import com.guerinet.mymartlet.util.manager.HomepageManager
import com.guerinet.mymartlet.util.prefs.DefaultTermPref
import com.guerinet.mymartlet.util.retrofit.TranscriptConverter.TranscriptResponse
import com.guerinet.mymartlet.util.room.daos.CourseDao
import com.guerinet.mymartlet.util.room.daos.TranscriptDao
import com.guerinet.suitcase.coroutines.bgDispatcher
import com.guerinet.suitcase.coroutines.uiDispatcher
import com.guerinet.suitcase.date.android.extensions.getLongDateString
import com.guerinet.suitcase.date.extensions.today
import com.guerinet.suitcase.settings.BooleanSetting
import com.guerinet.suitcase.util.extensions.getColorCompat
import com.guerinet.suitcase.util.extensions.openUrl
import com.guerinet.suitcase.util.extensions.toast
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.LocalDate
import kotlinx.datetime.TimeZone
import kotlinx.datetime.atTime
import kotlinx.datetime.plus
import kotlinx.datetime.toInstant
import kotlinx.datetime.until
import org.koin.android.ext.android.inject
import org.koin.core.qualifier.named
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import timber.log.Timber
import java.util.Locale
import java.util.Stack

/**
 * Displays the user's schedule
 * @author Julien Guerinet
 * @since 1.0.0
 */
class ScheduleActivity : DrawerActivity() {

    private val firstOpenPref by inject<BooleanSetting>(named(Prefs.IS_FIRST_OPEN))

    private val twentyFourHourPref by inject<BooleanSetting>(named(Prefs.SCHEDULE_24HR))

    private val defaultTermPref by inject<DefaultTermPref>()

    private val courseDao by inject<CourseDao>()

    private val transcriptDao by inject<TranscriptDao>()

    private val viewPager: ViewPager by lazy { findViewById(R.id.viewPager) }

    private var term: Term = defaultTermPref.term

    private val courses: MutableList<Course> = mutableListOf()

    // We need this to know which week to show in the landscape orientation
    private var date: LocalDate = LocalDate.today

    override val currentPage = HomepageManager.HomePage.SCHEDULE

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_schedule)

        val savedTerm = savedInstanceState?.get(Constants.TERM) as? Term
        if (savedTerm != null) {
            term = savedTerm
        }

        // Update the list of courses for this currentTerm and the starting date
        updateCoursesAndDate()

//        // Render the right view based on the orientation
//        if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
//            renderLandscapeView()
//        } else {
//            renderPortraitView()
//        }

        // Check if this is the first time the user is using the app
        if (firstOpenPref.value) {
            // Show them the walkthrough if it is
            start<WalkthroughActivity>(Prefs.IS_FIRST_OPEN to true)
            // Save the fact that the walkthrough has been seen at least once
            firstOpenPref.value = false
        }
    }

    // Only show the menu in portrait mode
    override fun onPrepareOptionsMenu(menu: Menu): Boolean =
        resources.configuration.orientation != Configuration.ORIENTATION_LANDSCAPE

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.refresh, menu)
        menuInflater.inflate(R.menu.change_semester, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_change_semester -> {
                TermDialogHelper(this, this, term) {
                    // If it's the same currentTerm as now, do nothing
                    if (it == term) {
                        return@TermDialogHelper
                    }

                    // Set the instance currentTerm
                    term = it

                    // Set the default currentTerm
                    defaultTermPref.term = term

                    // Update the courses
                    updateCoursesAndDate()

                    // Refresh the content
                    refreshCourses()
                }
                return true
            }
            R.id.action_refresh -> {
                refreshCourses()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        // Save the currentTerm
        outState.putSerializable(Constants.TERM, term)
    }

    private suspend fun updateCourses() {
        withContext(bgDispatcher) {
            // Clear the current courses
            courses.clear()

            // Get the new courses for the current currentTerm
            val localCourses = courseDao.getTermCourses(term)

            courses.addAll(localCourses)
        }
    }

    /**
     * Gets the courses for the given [Term]
     */
    private fun updateCoursesAndDate() {
        launch(Dispatchers.Default) {
            updateCourses()

            // Date is by default set to today
            date = LocalDate.today

            // Check if we are in the current semester
            if (term != Term.currentTerm()) {
                // If not, find the starting date of this semester instead of using today
                for (course in courses) {
                    if (course.startDate < date) {
                        date = course.startDate
                    }
                }
            }

            withContext(uiDispatcher) {
                // Title
                title = term.getString(this@ScheduleActivity)

                // TODO This only renders the portrait view
                renderPortraitView()
            }
        }
    }

    /**
     * Refreshes the list of courses for the given currentTerm and the user's transcript
     * (only available in portrait mode)
     */
    private fun refreshCourses() {
        if (!canRefresh()) {
            return
        }

        // Download the courses for this currentTerm
        mcGillService.schedule(term).enqueue(object : Callback<List<Course>> {
            override fun onResponse(call: Call<List<Course>>, response: Response<List<Course>>) {
                launch(Dispatchers.IO) {
                    courseDao.update(response.body() ?: listOf(), term)
                    withContext(uiDispatcher) {
                        // Update the view
                        toolbarProgress.isVisible = false
                        updateCourses()
                        @Suppress("PLUGIN_WARNING")
                        viewPager.adapter?.notifyDataSetChanged()
                    }
                }

                // Download the transcript (if ever the user has new semesters on their transcript)
                mcGillService.oldTranscript().enqueue(object : Callback<TranscriptResponse> {
                    override fun onResponse(
                        call: Call<TranscriptResponse>,
                        response: Response<TranscriptResponse>
                    ) {
                        launch(Dispatchers.IO) {
                            transcriptDao.update(response.body()?.transcript!!)
                        }
                    }

                    override fun onFailure(call: Call<TranscriptResponse>, t: Throwable) {
                        handleError("refreshing the transcript", t)
                    }
                })
            }

            override fun onFailure(call: Call<List<Course>>, t: Throwable) {
                handleError("refreshing the courses", t)
            }
        })
    }

    /* TODO Bring landscape view back
    /**
     * Renders the landscape view
     */
    @Suppress("PLUGIN_WARNING")
    private fun renderLandscapeView() {
        // Leave space at the top for the day names
        var dayView = View.inflate(this, R.layout.fragment_day_name, null)
        // Black line to separate the timetable from the schedule
        val dayViewLine = dayView.findViewById<View>(R.id.day_line)
        dayViewLine.visibility = View.VISIBLE

        // Add the day view to the top of the timetable
        timetableContainer.addView(dayView)

        // Find the index of the given date
        val currentDayIndex = date.dayOfWeek.value

        // Go through the 7 days of the week
        for (i in 1..7) {
            val day = DayOfWeek.of(i)

            // Set up the day name
            dayView = View.inflate(this, R.layout.fragment_day_name, null)
            val dayViewTitle = dayView.findViewById<TextView>(R.id.day_name)
            dayViewTitle.setText(DayUtils.getStringId(day))
            scheduleContainer!!.addView(dayView)

            // Set up the schedule container for that one day
            val scheduleContainer = LinearLayout(this)
            scheduleContainer.orientation = LinearLayout.VERTICAL
            scheduleContainer.layoutParams = LinearLayout.LayoutParams(
                    resources.getDimensionPixelSize(R.dimen.cell_landscape_width),
                    ViewGroup.LayoutParams.WRAP_CONTENT)

            // Fill the schedule for the current day
            fillSchedule(this.timetableContainer, scheduleContainer,
                    date.plusDays((i - currentDayIndex).toLong()), false)

            // Add the current day to the schedule container
            this.scheduleContainer!!.addView(scheduleContainer)

            // Line
            val line = View(this)
            line.setBackgroundColor(Color.BLACK)
            line.layoutParams = ViewGroup.LayoutParams(
                    resources.getDimensionPixelSize(R.dimen.schedule_line),
                    ViewGroup.LayoutParams.MATCH_PARENT)
            this.scheduleContainer!!.addView(line)
        }
    }
    */

    /**
     * Renders the portrait view
     */
    private fun renderPortraitView() {
        val adapter = ScheduleAdapter()

        // Set up the ViewPager
        viewPager.apply {
            this.adapter = adapter
            currentItem = adapter.startingDateIndex
            addOnPageChangeListener(object :
                androidx.viewpager.widget.ViewPager.OnPageChangeListener {
                override fun onPageScrolled(i: Int, v: Float, i2: Int) {}

                override fun onPageSelected(i: Int) {
                    //Update the date every time the page is turned to have the right
                    //  week if ever the user rotates his device
                    date = adapter.getDate(i)
                }

                override fun onPageScrollStateChanged(i: Int) {}
            })
        }
    }

    /**
     * Fills the schedule based on given data
     *
     * @param timetableContainer Container for the timetable
     * @param scheduleContainer Container for the schedule
     * @param date Date to fill the schedule for
     * @param clickable True if the user can click on the courses (portrait),
     * false otherwise (landscape)
     */
    private fun fillSchedule(
        timetableContainer: LinearLayout?,
        scheduleContainer: LinearLayout?,
        date: LocalDate,
        clickable: Boolean
    ) {
        if (timetableContainer == null || scheduleContainer == null) {
            return
        }

        // Clear everything out
        timetableContainer.removeAllViews()
        scheduleContainer.removeAllViews()

        // Go through the list of courses, find which ones are for the given date
        val courses = this.courses.filter { it.isForDate(date) }

        // TODO
//        // Set up the DateTimeFormatter we're going to use for the hours
//        val pattern = if (twentyFourHourPref.value) "HH:mm" else "hh a"
//        val formatter = DateTimeFormatter.ofPattern(pattern)

        // This will be used of an end time of a course when it is added to the schedule container
        var currentCourseEndTime: LocalTime? = null

        // Cycle through the hours
        for (hour in 8..21) {
            // Start inflating a timetable cell
            val timetableCell = View.inflate(this, R.layout.item_day_timetable, null)

            // Put the correct time
            val time = timetableCell.findViewById<TextView>(R.id.cell_time)
            val localDateTime = LocalDate.today.atTime(hour, 0)
            time.text = localDateTime.time.getShortTimeString()

            // Add it to the right container
            timetableContainer.addView(timetableCell)

            // Cycle through the half hours
            var min = 0
            while (min < 31) {
                // Initialize the current course to null
                var currentCourse: Course? = null

                // Get the current time
                val currentTime = LocalTime(hour, min)

                // if currentCourseEndTime = null (no course is being added) or it is equal to
                //  the current time in min (end of a course being added) we need to add a new view
                if (currentCourseEndTime == null || currentCourseEndTime == currentTime) {
                    // Reset currentCourseEndTime
                    currentCourseEndTime = null

                    // Check if there is a course at this time
                    for (course in courses) {
                        // If there is, set the current course to that time, and calculate the
                        //  ending time of this course
                        if (course.roundedStartTime == currentTime) {
                            currentCourse = course
                            currentCourseEndTime = course.roundedEndTime
                            break
                        }
                    }

                    val scheduleCell: View

                    // There is a course at this time
                    if (currentCourse != null) {
                        // Inflate the right view
                        scheduleCell = View.inflate(this, R.layout.item_day_class, null)

                        // Set up all of the info
                        val code = scheduleCell.findViewById<TextView>(R.id.code)
                        code.text = currentCourse.code

                        val type = scheduleCell.findViewById<TextView>(R.id.type)
                        type.text = currentCourse.type

                        val courseTime = scheduleCell.findViewById<TextView>(R.id.course_time)
                        courseTime.text = currentCourse.timeString

                        val location = scheduleCell.findViewById<TextView>(R.id.course_location)
                        location.text = currentCourse.location

                        // Find out how long this course is in terms of blocks of 30 min
                        val start = LocalDate
                            .today
                            .atTime(currentCourse.roundedStartTime)
                            .toInstant(TimeZone.UTC)
                        val end = LocalDate
                            .today
                            .atTime(currentCourse.roundedEndTime)
                            .toInstant(TimeZone.UTC)

                        val length = start.until(end, DateTimeUnit.MINUTE).toInt() / 30

                        // Set the height of the view depending on this height
                        val lp = LinearLayout.LayoutParams(
                            LinearLayout.LayoutParams.MATCH_PARENT,
                            resources.getDimension(R.dimen.cell_30min_height).toInt() * length
                        )
                        scheduleCell.layoutParams = lp

                        // Check if we need to make the course clickable
                        if (clickable) {
                            // We need a final variable for the onClick listener
                            val course = currentCourse
                            // OnClick: CourseActivity (for a detailed description of the course)
                            scheduleCell.setOnClickListener { showCourseDialog(course) }
                        } else {
                            scheduleCell.isClickable = false
                        }
                    } else {
                        // Inflate the empty view
                        scheduleCell = View.inflate(this, R.layout.item_day_empty, null)
                    }

                    // Add the given view to the schedule container
                    scheduleContainer.addView(scheduleCell)
                }
                min += 30
            }
        }
    }

    /**
     * Shows a dialog with course information
     *
     * @param course Clicked [Course]
     */
    private fun showCourseDialog(course: Course) {
        analytics.event("schedule_course_details")

        // Set up the view in the dialog
        val view = ScrollView(this)
        val container = LinearLayout(this)
        container.orientation = LinearLayout.VERTICAL
        view.addView(container)

        // Create the dialog
        val alert = AlertDialog.Builder(this)
            .setView(view)
            .setCancelable(true)
            .setNeutralButton(R.string.done) { dialog, _ -> dialog.dismiss() }
            .show()

        // Populate the form
        val shape = Morf.shape
        //                .setShowLine(false)
        //                .setInputDefaultBackground(android.R.color.transparent)
        container.morf(shape) {
            // Code
            addTextInput(R.string.course_code, course.code)

            // Title
            addTextInput(R.string.course_name, course.title)

            // Time
            addTextInput(R.string.course_time_title, course.timeString)

            // Location
            addTextInput(R.string.course_location, course.location)

            // Type
            addTextInput(R.string.schedule_type, course.type)

            // Instructor
            addTextInput(R.string.course_prof, course.instructor)

            // Section
            addTextInput(R.string.course_section, course.section)

            // Credits
            addTextInput(R.string.course_credits_title, course.credits.toString())

            // CRN
            addTextInput(R.string.course_crn, course.crn.toString())

            val color = getColorCompat(R.color.red)

            // Docuum
            borderlessButton {
                layoutParams(
                    LinearLayout.LayoutParams(
                        ViewGroup.LayoutParams.WRAP_CONTENT,
                        ViewGroup.LayoutParams.WRAP_CONTENT
                    ), Gravity.CENTER
                )
                textId = R.string.course_docuum
                textColor = color
                onClick {
                    openUrl(
                        "http://www.docuum.com/mcgill/${course.subject.lowercase(Locale.getDefault())}" +
                                "/${course.number}"
                    )
                }
            }

            // Maps
            borderlessButton {
                layoutParams(
                    LinearLayout.LayoutParams(
                        ViewGroup.LayoutParams.WRAP_CONTENT,
                        ViewGroup.LayoutParams.WRAP_CONTENT
                    ), Gravity.CENTER
                )
                textId = R.string.course_map
                textColor = color
                onClick {
                    // Try to find a place that has the right name
                    launch(uiDispatcher) {
                        // Load the places from the Firestore
                        val place =
                            Place.loadPlaces().firstOrNull { course.location.contains(it.coursePlaceName, true) }

                        if (place == null) {
                            // Tell the user
                            toast(getString(R.string.error_place_not_found, course.location))
                            // Send a Crashlytics report
                            Timber.e(NullPointerException("Location not found: ${course.location}"))
                        } else {
                            // Close the dialog
                            alert.dismiss()
                            // Open the map to the given place
                            val intent = intentFor<MapActivity>(Constants.ID to place.id)
                            switchDrawerActivity(intent)
                        }
                    }
                }
            }
        }
    }

    private fun Morf.addTextInput(@StringRes hintId: Int, text: String) = textInput {
        this.hintId = hintId
        this.text = text
        isEnabled = false
    }

    /**
     * Adapter used for the ViewPager in the portrait view of the schedule
     */
    internal inner class ScheduleAdapter : androidx.viewpager.widget.PagerAdapter() {

        /**
         * The initial date to use as a reference
         */
        private val startingDate = date
        /**
         * The index of the starting date (offset of 500001 to get the right day)
         */
        val startingDateIndex = 500001 + date.dayOfWeek.value

        private val holders = Stack<DayHolder>()

        override fun instantiateItem(collection: ViewGroup, position: Int): Any {
            val holder = if (holders.isNotEmpty()) {
                holders.pop()
            } else {
                DayHolder(
                    LayoutInflater.from(collection.context)
                        .inflate(R.layout.fragment_day, collection, false)
                )
            }

            // Get the date for this view and set it
            holder.bind(getDate(position))

            collection.addView(holder.view)
            return holder.view
        }

        override fun destroyItem(collection: ViewGroup, position: Int, view: Any) {
            val dayView = view as? View ?: error("PagerAdapter item was not View type")
            collection.removeView(dayView)
            holders.push(DayHolder(view))
        }

        override fun getCount() = 1000000

        fun getDate(position: Int): LocalDate =
            startingDate.plus(position - startingDateIndex, DateTimeUnit.DAY)

        // This is to force the refreshing of all of the views when the view is reloaded
        override fun getItemPosition(`object`: Any): Int =
            androidx.viewpager.widget.PagerAdapter.POSITION_NONE

        override fun isViewFromObject(view: View, `object`: Any): Boolean = view == `object`

        inner class DayHolder(val view: View) {

            private val dayTitle by lazy<TextView> { view.findViewById(R.id.dayTitle) }
            private val dayDate by lazy<TextView> { view.findViewById(R.id.dayDate) }
            private val timetableContainer by lazy<LinearLayout> {
                view.findViewById(R.id.timetableContainer)
            }
            private val scheduleContainer by lazy<LinearLayout> {
                view.findViewById(R.id.scheduleContainer)
            }

            internal fun bind(date: LocalDate) {
                // Set the titles
                view.apply {
                    dayTitle.setText(DayUtils.getStringId(date.dayOfWeek))
                    dayDate.text = date.getLongDateString()

                    // Fill the schedule up
                    fillSchedule(timetableContainer, scheduleContainer, date, true)
                }
            }
        }
    }
}
